### PR TITLE
[persist] Minor fixups

### PR DESCRIPTION
**Pull request by bkirwi**
Primarily providing more-efficient iterators over batches, plus some other odds & ends.

### Motivation

Previously-unspecified refactoring. (I think most of these are valuable on their own, but efficiently iterating over spine batches looks like it will be useful for some upcoming changes.)

### Tips for reviewer

Unsolicited refactoring PR, so feel free to push back on whatever. In particular, I wasn't sure how far to go with migrating `map_batches`; all of the usages can be easily ported to a for loop, so I just picked the ones that either seemed simpler with iterator methods or got to save a little work.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/internal/trace.rs [hunk 1]
```diff
@@ -984,17 +984,18 @@ struct Spine<T> {
...
-                }
-                MergeState::Single(batch) => f(batch),
-                _ => {}
-            }
+        for batch in self.spine_batches() {
+            f(batch);
         }
     }
+
+    pub fn spine_batches(&self) -> impl Iterator<Item = &SpineBatch<T>> {
```
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs:974]: Clever!

File: src/persist-client/src/internal/trace.rs [hunk 2]
```diff
@@ -369,22 +369,19 @@ impl<T> Trace<T> {
...
-    pub fn batches(&self) -> impl IntoIterator<Item = &HollowBatch<T>> {
-        // It should be possible to do this without the Vec.
-        let mut batches = Vec::new();
-        self.map_batches(|b| batches.push(b));
-        batches
+    pub fn batches(&self) -> impl Iterator<Item = &HollowBatch<T>> {
+        self.spine
+            .spine_batches()
+            .flat_map(|b| match b {
+                SpineBatch::Merged(b) => std::slice::from_ref(b),
```
Comment (danhhz) [inline src/persist-client/src/internal/trace.rs:377]: TIL about `std::slice::from_ref`

**Review by danhhz — APPROVED** [review]
Nice!

> I wasn't sure how far to go with migrating map_batches

I'm fine migrating all of them and removing it if you like. I prefer the iterator over map_batches, but didn't have the two insights you had here to unlock doing it in a performant way. But also feel free to punt and merge as-is

File: src/persist-client/src/internal/trace.rs (hunk 2 above)
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:377]: Unironically one of my favourite rust functions. 😬 Comes up surprisingly often!

File: src/persist-client/src/internal/trace.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/internal/trace.rs:974]: I stole this trick from `MfpPlan::evaluate`, so credit goes to Frank I think!

**Comment by bkirwi** [comment]
Thanks! I may eventually follow up with the rest of them, but I'll get this in for now.

**Pull request merged**